Feature/quadrangle mesh

Mateasek opened the pull request.

Mateasek committed:
add b2 quadrangles and quadrange mapping

Quadrangle indicies are kept and provided as property. Mapping from the
quadrangles to the b2 grid is also kept.

changed plot_mesh to plot_triangle_mesh, added plot_quadrangle_mesh
changed some plot properties

Mateasek committed:
add quadrangle mesh plot demo

Mateasek commented:
@vsnever  could you please look at this PR? Thanks.

vsnever commented on cherab/solps/mesh_geometry.py:
```diff
@@ -134,6 +134,19 @@ def __init__(self, r, z, vol, neighbix, neighbiy):
         self._triangle_to_grid_map[::2, 1] = xm.flatten()
         self._triangle_to_grid_map[1::2] = self._triangle_to_grid_map[::2]
 
+        # add quadrangle b2 grid
+        self._quadrangles = np.zeros((ng, 4), dtype=np.int32)
+        self._quadrangles[:, 0] = unique_vertices[0:ng]
+        self._quadrangles[:, 1] = unique_vertices[2 * ng: 3 * ng]
+        self._quadrangles[:, 2] = unique_vertices[3 * ng: 4 * ng]
+        self._quadrangles[:, 3] = unique_vertices[ng: 2 * ng]
+        
+        # add mapping from quadrangles to the b2 grid
+        self._quadrangle_to_grid_map = np.zeros((ng, 2), dtype=np.int32)
+        self._quadrangle_to_grid_map[:, 0] = ym.flatten()
+        self._quadrangle_to_grid_map[:, 1] = xm.flatten()
```
I think it will be clearer to define the quadrangle grid first and then use it to create the triangle grid:
```python
        ng = self._nx * self._ny  # grid size
        # add quadrangle b2 grid
        self._quadrangles = np.zeros((ng, 4), dtype=np.int32)
        self._quadrangles[:, 0] = unique_vertices[0:ng]
        self._quadrangles[:, 1] = unique_vertices[2 * ng: 3 * ng]
        self._quadrangles[:, 2] = unique_vertices[3 * ng: 4 * ng]
        self._quadrangles[:, 3] = unique_vertices[ng: 2 * ng]

        # add mapping from quadrangles to the b2 grid
        ym, xm = np.meshgrid(np.arange(self._ny, dtype=np.int32), np.arange(self._nx, dtype=np.int32), indexing='ij')
        self._quadrangle_to_grid_map = np.zeros((ng, 2), dtype=np.int32)
        self._quadrangle_to_grid_map[:, 0] = ym.flatten()
        self._quadrangle_to_grid_map[:, 1] = xm.flatten()

        # Number of triangles must be equal to number of rectangle centre points times 2.
        self._num_tris = ng * 2
        self._triangles = np.zeros((self._num_tris, 3), dtype=np.int32)
        self._triangles[0::2, 0] = self._quadrangles[:, 0]
        self._triangles[0::2, 1] = self._quadrangles[:, 1]
        self._triangles[0::2, 2] = self._quadrangles[:, 2]
        # Split the quad cell into two triangular cells.
        self._triangles[1::2, 0] = self._quadrangles[:, 2]
        self._triangles[1::2, 1] = self._quadrangles[:, 3]
        self._triangles[1::2, 2] = self._quadrangles[:, 0]

        # Each triangle cell is mapped to the tuple ID (ix, iy) of its parent mesh cell.
        self._triangle_to_grid_map = np.zeros((self._num_tris, 2), dtype=np.int32)
        self._triangle_to_grid_map[::2, 0] = ym.flatten()
        self._triangle_to_grid_map[::2, 1] = xm.flatten()
        self._triangle_to_grid_map[1::2] = self._triangle_to_grid_map[::2]
```
Because now we can see how the quadrangles are triangulated.

Mateasek replied:
Good point, I didn't want to change code which was already there, but this is indeed more instructive. I applied the proposed changes

vsnever commented on cherab/solps/mesh_geometry.py:
```diff
@@ -302,6 +329,26 @@ def plot_mesh(self):
+        ax.set_aspect(1)
+        ax.set_xlim(self.mesh_extent["minr"], self.mesh_extent["maxr"])
+        ax.set_ylim(self.mesh_extent["minz"], self.mesh_extent["maxz"])
 
```
These two function can be simplified by using `PolyCollection` instead of `PatchCollection` but also expanded in functionality by introducing additional argument `solps_data`. I propose the following (optional) change:
```python
    def plot_triangle_mesh(self, solps_data=None):
        """
        Plot the triangle mesh grid geometry to a matplotlib figure.

        :param solps_data: Data array defined on the SOLPS mesh
        """
        fig, ax = plt.subplots()
        verts = self.vertex_coordinates[self.triangles]
        if solps_data is None:
            collection_mesh = PolyCollection(verts, facecolor="none", edgecolor='b', linewidth=0.5)
        else:
            collection_mesh = PolyCollection(verts)
            collection_mesh.set_array(solps_data[self.triangle_to_grid_map[:, 0], self.triangle_to_grid_map[:, 1]])
        ax.add_collection(collection_mesh)
        ax.set_aspect(1)
        ax.set_xlim(self.mesh_extent["minr"], self.mesh_extent["maxr"])
        ax.set_ylim(self.mesh_extent["minz"], self.mesh_extent["maxz"])
        ax.set_ylabel("R [m]")
        ax.set_ylabel("Z [m]")

        return ax

    def plot_quadrangle_mesh(self, solps_data=None):
        """
        Plot the quadrangle mesh grid geometry to a matplotlib figure.

        :param solps_data: Data array defined on the SOLPS mesh
        """
        fig, ax = plt.subplots()
        verts = self.vertex_coordinates[self.quadrangles]
        if solps_data is None:
            collection_mesh = PolyCollection(verts, facecolor="none", edgecolor='b', linewidth=0.5)
        else:
            collection_mesh = PolyCollection(verts)
            collection_mesh.set_array(solps_data[self.quadrangle_to_grid_map[:, 0], self.quadrangle_to_grid_map[:, 1]])
        ax.add_collection(collection_mesh)
        ax.set_aspect(1)
        ax.set_xlim(self.mesh_extent["minr"], self.mesh_extent["maxr"])
        ax.set_ylim(self.mesh_extent["minz"], self.mesh_extent["maxz"])
        ax.set_ylabel("R [m]")
        ax.set_ylabel("Z [m]")

        return ax
```

Mateasek replied:
This is a good point. I applied the proposed changes and also added the ax as an optional parameter, so the methods can be used to add the mesh to externally created plots.

vsnever commented on demos/plots/quadrangle_mesh.py:
```diff
@@ -0,0 +1,60 @@
+print('Loading simulation...')
+sim = load_solps_from_raw_output(simulation_directory)
+
+#get the neccessary data from SOLPSSimulation
+vertices = sim.mesh.vertex_coordinates
+quadrangles = sim.mesh.quadrangles
+mesh_extent = sim.mesh.mesh_extent
+quadrangle_to_grid_map = sim.mesh.quadrangle_to_grid_map
+
+#plot the quadrangle b2 mesh
+collection_mesh = PolyCollection(vertices[quadrangles], facecolor="none")
+
+fig_mesh, ax = plt.subplots(figsize=(10, 20))
+ax.add_collection(collection_mesh)
+ax.set_xlim(mesh_extent["minr"], mesh_extent["maxr"])
+ax.set_ylim(mesh_extent["minz"], mesh_extent["maxz"])
+ax.set_aspect(1)
+ax.set_ylabel("R [m]")
+ax.set_ylabel("Z [m]")
+ax.set_title("Quadrangle B2 Mesh")
+
+#plot the quadrangle b2 mesh with b2 ion temperature values
+collection_values = PolyCollection(vertices[quadrangles])
+collection_values.set_array(sim.ion_temperature[quadrangle_to_grid_map[:,0], quadrangle_to_grid_map[:,1]])
+
+fig_mesh, ax = plt.subplots(figsize=(10, 20))
+ax.add_collection(collection_values)
+ax.set_xlim(mesh_extent["minr"], mesh_extent["maxr"])
+ax.set_ylim(mesh_extent["minz"], mesh_extent["maxz"])
+ax.set_aspect(1)
+ax.set_ylabel("R [m]")
+ax.set_ylabel("Z [m]")
+ax.set_title("B2 Ion Temperature")
```
If the `plot_quadrangle_mesh()` is updated as shown above, then this code can be replaced with this:
```python
#plot the quadrangle b2 mesh
ax = sim.mesh.plot_quadrangle_mesh()
ax.set_title("Quadrangle B2 Mesh")
ax.get_figure().set_size_inches((10, 20))

#plot the quadrangle b2 mesh with b2 ion temperature values
ax = sim.mesh.plot_quadrangle_mesh(solps_data=sim.ion_temperature)
ax.get_figure().colorbar(ax.collections[0], aspect=40)
ax.get_figure().set_size_inches((10, 20))
ax.set_title("B2 Ion Temperature [eV]")
plt.show()
```
Also, please don't forget to add `plt.show()` to the end of all plotting demos so that they can be run correctly without Jupyter.

Mateasek replied:
I adjusted the plot demo to show how can the plot methods be used and also left the example of what is done within the methods. I also renamed the file from quadrangle_mesh.py to mesh_and_values.py, since there are now both examples of triangular and quadrangle mesh plots.

vsnever reviewed (changes requested):
Hi @Mateasek,

This feature makes the plotting much easier, thank you. I have a few optional suggestions for improving the code.

Also, please replace `mesh.plot_mesh()` with `mesh.plot_triangle_mesh()` or `mesh.plot_quadrangle_mesh()` in these demos: generomak_raw_files_plot_profiles.py, mastu_solps_from_mdsplus.py and solps_from_balance.py.

Mateasek committed:
Change mesh generation and mesh plotting methods

mesh generation changed to improve the readability of trangulation of
the b2 quadrangle grid.
plot_quadrangle_mesh and plot_triangle_mesh now can plot mesh values and
accept external axes.

Mateasek committed:
change plot demo to react to changes

shows how to use the plot_quadrangle_mesh and plot_triangle_mesh

Mateasek committed:
rename the plot mesh demo

Mateasek commented:
Hi @vsnever , thank you very much for the review and suggestions! I applied the changes

vsnever reviewed (approved):
Great, thanks! I think this can be merged.

jacklovell commented:
Thanks Matej, and thanks Vlad for the review!

vsnever commented:
In the review I mentioned that all `plot_mesh()` calls in the demos should be replaced with `plot_triangle_mesh()` or `plot_quadrangle_mesh()` calls as `plot_mesh()` is no longer available. However, I completely forgot about this when I approved the changes the next day, sorry.

Mateasek commented:
Me too, I will make a separate PR to fix it. Thanks for remembering.
